File: Code/Chapter 9/C#/Exercise 9.4/Form1.cs
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using System.Data.SqlClient;

#endregion

namespace Exercise_9_4
{
   partial class Form1 : Form
   {
      private string connectionString = "Server=(local);Database=Test;Integrated Security=SSPI";

      private DataTable animalsTable;

      public Form1()
      {
         InitializeComponent();

         // Setup the schema for the Table.
         animalsTable = new DataTable("Animals");
         DataColumn dc ;

         dc = new DataColumn("AnimalID");
         dc.Unique = true;
         dc.AutoIncrement = true;
         dc.AutoIncrementSeed = -1;
         dc.AutoIncrementStep = -1;
         dc.DataType = typeof(System.Int32);
         animalsTable.Columns.Add(dc);

         dc = new DataColumn("AnimalName");
         animalsTable.Columns.Add(dc);

         dc = new DataColumn("AccountNumber");
         dc.ReadOnly = true;
         animalsTable.Columns.Add(dc);

         // DataBind it, even though it has no rows in it yet.
         dgView.DataSource = animalsTable;
      }

      private void btnLoad_Click(object sender, EventArgs e)
      {
         SqlDataAdapter sqlDA = new SqlDataAdapter("Select * from Animals", connectionString);
         animalsTable.Rows.Clear();
         sqlDA.Fill(animalsTable);
      }

      private void btnSave_Click(object sender, EventArgs e)
      {
         SqlCommand insertCommand = new SqlCommand();
         insertCommand.CommandType = CommandType.StoredProcedure;
         
         SqlParameter param;

         param = new SqlParameter("@AnimalID", SqlDbType.Int);
         param.Direction = ParameterDirection.Output;
         param.SourceColumn = "AnimalID";
         insertCommand.Parameters.Add(param);

         param = new SqlParameter("@AnimalName", SqlDbType.VarChar);
         param.SourceColumn = "AnimalName";
         insertCommand.Parameters.Add(param);

         param = new SqlParameter("@AccountNumber", SqlDbType.Int);
         param.SourceColumn = "AccountNumber";
         param.Size = 40;
         param.Direction = ParameterDirection.Output;
         insertCommand.Parameters.Add(param);

         insertCommand.CommandText = "UP_ANIMALINSERT";
         SqlDataAdapter sqlDA = new SqlDataAdapter("Select * from Animals", connectionString);
         insertCommand.Connection = new SqlConnection(connectionString);
         insertCommand.UpdatedRowSource = UpdateRowSource.Both;
         sqlDA.InsertCommand = insertCommand;

         sqlDA.Update(animalsTable);
      }
   }
}

File: Code/Chapter 9/VB.NET/Exercise 9.4/Form1.vb
Imports System.Data
Imports System.Data.SqlClient

Public Class Form1
   Private connectionString As String = "Server=(local);Database=Test;Integrated Security=SSPI"
   Private animalsTable As DataTable

   Private Sub btnLoad_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnLoad.Click
      Dim sqlDA As SqlDataAdapter = New SqlDataAdapter("Select * from Animals", connectionString)
      animalsTable.Rows.Clear()
      sqlDA.Fill(animalsTable)
   End Sub
   Private Sub btnSave_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSave.Click
      Dim insertCommand As SqlCommand = New SqlCommand()
      insertCommand.CommandType = CommandType.StoredProcedure

      Dim param As SqlParameter

      param = New SqlParameter("@AnimalID", SqlDbType.Int)
      param.Direction = ParameterDirection.Output
      param.SourceColumn = "AnimalID"
      insertCommand.Parameters.Add(param)

      param = New SqlParameter("@AnimalName", SqlDbType.VarChar)
      param.SourceColumn = "AnimalName"
      insertCommand.Parameters.Add(param)

      param = New SqlParameter("@AccountNumber", SqlDbType.Int)
      param.SourceColumn = "AccountNumber"
      param.Size = 40
      param.Direction = ParameterDirection.Output
      insertCommand.Parameters.Add(param)

      insertCommand.CommandText = "UP_ANIMALINSERT"
      Dim sqlDA As SqlDataAdapter = New SqlDataAdapter("Select * from Animals", connectionString)
      insertCommand.Connection = New SqlConnection(connectionString)
      insertCommand.UpdatedRowSource = UpdateRowSource.Both
      sqlDA.InsertCommand = insertCommand

      sqlDA.Update(animalsTable)
   End Sub

   Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
      ' Setup the schema for the Table.
      animalsTable = New DataTable("Animals")
      Dim dc As DataColumn

      dc = New DataColumn("AnimalID")
      dc.Unique = True
      dc.AutoIncrement = True
      dc.AutoIncrementSeed = -1
      dc.AutoIncrementStep = -1
      dc.DataType = Type.GetType("System.Int32") 
      animalsTable.Columns.Add(dc)

      dc = New DataColumn("AnimalName")
      animalsTable.Columns.Add(dc)

      dc = New DataColumn("AccountNumber")
      dc.ReadOnly = True
      animalsTable.Columns.Add(dc)

      ' DataBind it, even though it has no rows in it yet.
      dgView.DataSource = animalsTable
   End Sub
End Class
